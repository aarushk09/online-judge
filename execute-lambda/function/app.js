const { exec, execFile, execFileSync } = require('child_process');
const { writeFileSync, readFileSync, existsSync, mkdirSync, unlinkSync, rmdirSync } = require('fs');

exports.handler = function (event, context, callback) {
  if (event.type === "compile") {
    if (!existsSync('/tmp/out')) mkdirSync('/tmp/out');
    writeFileSync(`/tmp/out/${event.filename}`, event.sourceCode);

    if (!["cpp", "java"].includes(event.language)) {
      callback("Unknown Language");
      return;
    }

    const command = {
      'cpp': "g++",
      'java': "javac",
    }[event.language];

    const compile = execFile(
      command,
      [
        ...event.compilerOptions.split(" "),
        ...(event.language === 'cpp' ? ['-o', '/tmp/out/prog'] : ['-d', 'out']),
        `/tmp/out/${event.filename}`
      ].filter(x => !!x),
      (error, stdout, stderr) => {
        if (error) {
          callback(error);
        }

        writeFileSync("/tmp/out/run.sh", event.language === "cpp" ? "./prog" : "java \"" + event.filename.split(".")[0] + "\"");
        execFileSync('zip', ['-r', '/tmp/out.zip', '-j', '/tmp/out']);

        const base64Output = readFileSync('/tmp/out.zip', { encoding: 'base64' });

        callback(null, {
          status: "success",
          output: base64Output
        });

        unlinkSync('/tmp/out.zip');
        rmdirSync('/tmp/out', { recursive: true });
      }
    );
  } else if (event.type === "execute") {
    if (!existsSync('/tmp/out')) mkdirSync('/tmp/out');
    writeFileSync("/tmp/program.zip", event.payload, "base64");
    writeFileSync("/tmp/input.txt", event.input);
    execFileSync("unzip", ["-o", "/tmp/program.zip", "-d", "/tmp/program"]);
    exec("cd /tmp/program; sh /tmp/program/run.sh < /tmp/input.txt", (error, stdout, stderr) => {
      if (error) {
        callback(error);
      }
      callback(null, {
        status: "success",
        stdout,
        stderr,
      });
      unlinkSync('/tmp/program.zip');
      unlinkSync('/tmp/input.txt');
      rmdirSync('/tmp/program', { recursive: true });
      rmdirSync('/tmp/out', { recursive: true });
    });
  }
}