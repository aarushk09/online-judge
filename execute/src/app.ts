import { spawnSync, execFileSync, spawn, execSync } from "child_process";
import {
  writeFileSync,
  existsSync,
  mkdirSync,
  unlinkSync,
  rmdirSync,
  readFileSync,
  rmSync,
} from "fs";
import {
  ExecuteProcessOutput,
  parseReturnInfoOfSpawn,
  zipAndRemoveOutDir,
} from "./utils";

export type ExecuteEvent =
  | {
      type: "compile";
      filename: string;
      sourceCode: string;
      compilerOptions: string;
      language: "cpp" | "java" | "py";
    }
  | {
      type: "execute";
      payload: string;
      input: string;
      /**
       * If provided, in addition to standard input/output,
       * file I/O will also be used. [fileIOName].in will
       * be written and [fileIOName].out will be read.
       */
      fileIOName?: string;
      timeout?: number;
    };

export type CompilationResult =
  | {
      status: "internal_error";
      /**
       * Ex. "Unknown Language"
       */
      message: string;
    }
  | {
      status: "compile_error";
      message: string;
    }
  | {
      status: "success";
      /**
       * Base64 encoded ZIP file containing the output
       */
      output: string;
      processOutput: ExecuteProcessOutput | null;
    };

export type ExecutionResult =
  | {
      status: "success";
    } & ExecuteProcessOutput;

const MAX_BUFFER_SIZE = 1024 * 1024 * 30; // 30mb. note that 6mb is API gateway limit for lambda

export const lambdaHandler = async function (
  event: ExecuteEvent
): Promise<CompilationResult | ExecutionResult> {
  const IS_AWS = process.env.AWS_EXECUTION_ENV;
  // AWS Lambda reuses containers between executions.
  // Occasionally, this causes ENOSPC no space left on device errors
  // This shouldn't happen because we clean up /tmp/program and /tmp/out,
  // but maybe a malicious user is creating many large files in /tmp or something.
  // Anyway, as a dirty fix, we can just clear the /tmp directory every time.
  // However, we only want to do this on AWS lambda, not anywhere else
  // (ie. when testing locally we probably don't want to delete /tmp)
  // See https://github.com/cpinitiative/online-judge/issues/13
  if (IS_AWS) {
    if (existsSync("/tmp")) {
      execSync("rm -rf /tmp/..?* /tmp/.[!.]* /tmp/*");
    }
  } else {
    if (existsSync("/tmp/out")) {
      rmdirSync("/tmp/out", { recursive: true });
    }
    if (existsSync("/tmp/program")) {
      rmdirSync("/tmp/program", { recursive: true });
    }
  }
  mkdirSync("/tmp/out", { recursive: true });

  if (event.type === "compile") {
    writeFileSync(`/tmp/out/${event.filename}`, event.sourceCode);

    if (!["cpp", "java", "py"].includes(event.language)) {
      return {
        status: "internal_error",
        message: "Unknown Language: " + event.language,
      };
    }

    if (event.language === "py") {
      writeFileSync("/tmp/out/run.sh", 'python3.8 "' + event.filename + '"');

      return {
        status: "success",
        output: zipAndRemoveOutDir(),
        processOutput: null,
      };
    }

    // if process.env.AWS_EXECUTION_ENV is set, then we're running in Amazon Linux 2, which has a special path to g++
    const cppCompilationCommand = IS_AWS
        ? "/opt/rh/devtoolset-10/root/usr/bin/g++"
        : "g++",
      javaCompilationCommand = "javac";
    const cppCompilationOptions = ["-o", "/tmp/out/prog"],
      javaCompilationOptions = ["-d", "/tmp/out"];
    const cppExecutionCommand = "./prog",
      javaExecutionCommand = 'java "' + event.filename.split(".")[0] + '"';

    const spawnResult = spawnSync(
      event.language === "cpp" ? cppCompilationCommand : javaCompilationCommand,
      [
        ...event.compilerOptions.split(" "),
        ...(event.language === "cpp"
          ? cppCompilationOptions
          : javaCompilationOptions),
        event.filename,
      ].filter((x) => !!x),
      {
        cwd: "/tmp/out",
        timeout: 6000,
        shell: true,
      }
    );
    const processOutput = parseReturnInfoOfSpawn(spawnResult);
    if (spawnResult.status !== 0) {
      return {
        status: "compile_error",
        message: processOutput.stderr ?? "",
      };
    }

    writeFileSync(
      "/tmp/out/run.sh",
      event.language === "cpp" ? cppExecutionCommand : javaExecutionCommand
    );

    return {
      status: "success",
      output: zipAndRemoveOutDir(),
      processOutput: processOutput,
    };
  } else if (event.type === "execute") {
    writeFileSync("/tmp/program.zip", event.payload, "base64");
    execFileSync("unzip", ["-o", "/tmp/program.zip", "-d", "/tmp/program"]);

    if (event.fileIOName) {
      writeFileSync(`/tmp/program/${event.fileIOName}.in`, event.input);
    }

    const spawnResult = spawnSync(
      `ulimit -c 0 && ulimit -s unlimited && /usr/bin/time -v /usr/bin/timeout ${(
        (event.timeout ?? 5000) / 1000
      ).toFixed(3)}s sh /tmp/program/run.sh`,
      {
        cwd: "/tmp/program",
        input: event.input,
        shell: true,
        maxBuffer: MAX_BUFFER_SIZE,
      }
    );

    let fileIOInfo: Partial<ExecuteProcessOutput> = {};
    if (event.fileIOName) {
      if (existsSync(`/tmp/program/${event.fileIOName}.out`)) {
        fileIOInfo.fileOutput = readFileSync(
          `/tmp/program/${event.fileIOName}.out`
        ).toString();
      } else {
        fileIOInfo.fileOutput = null;
      }
    }

    unlinkSync("/tmp/program.zip");
    rmdirSync("/tmp/program", { recursive: true });
    rmdirSync("/tmp/out", { recursive: true });

    return {
      status: "success",
      ...parseReturnInfoOfSpawn(spawnResult),
      ...fileIOInfo,
    };
  } else {
    return {
      status: "internal_error",
      // @ts-expect-error
      message: "Unknown event type: " + event.type,
    };
  }
};
