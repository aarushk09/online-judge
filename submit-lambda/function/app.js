const aws = require("aws-sdk");
const { v4: uuidv4 } = require('uuid');

const lambda = new aws.Lambda({
  region: "us-west-1"
});
const docClient = new aws.DynamoDB.DocumentClient();

exports.handler = function (event, context, callback) {
  const submissionId = uuidv4();
  lambda.invoke({
    FunctionName: 'execute',
    Payload: JSON.stringify({
      type: "compile",
      language: event.language,
      compilerOptions: event.compilerOptions,
      filename: event.filename,
      sourceCode: event.sourceCode,
    }, null, 2)
  }, function(error, data) {
    if (error) {
      context.done('error', error);
    }
    if(data.Payload){
      lambda.invoke({
        FunctionName: 'execute',
        Payload: JSON.stringify({
          type: "execute",
          payload: JSON.parse(data.Payload.toString()).output,
          input: event.input
        }, null, 2)
      }, function(error, data) {
        if (error) {
          context.done('error', error);
        }
        if(data.Payload){
          context.succeed(JSON.parse(data.Payload));
        }
      });
    }
  });
}
