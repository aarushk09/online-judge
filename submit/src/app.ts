import { APIGatewayProxyEvent, APIGatewayProxyResult} from "aws-lambda";
import { buildResponse, extractTimingInfo } from "./utils";
import { S3Client, AbortMultipartUploadCommand, ListObjectsCommand, GetObjectCommand } from "@aws-sdk/client-s3";
import { LambdaClient, InvokeCommand } from "@aws-sdk/client-lambda";
const { DynamoDBClient, ListTablesCommand, CreateTableCommand, CreateTableCommandInput, PutItemCommand, GetItemCommand } = require("@aws-sdk/client-dynamodb");
import { v4 as uuidv4 } from 'uuid';


const client = new LambdaClient({
  region: "us-west-1",
});

const config = {
  accessKeyId: "AKIAWCOXREPQGDVKJOOV",
  secretAccessKey: "AKIAWCOXREPQGDVKJOOV",
};

  const db = new DynamoDBClient({credentials: config,   region: "us-west-1"});
  const s3 = new S3Client({credentials: config,   region: "us-west-1"});

// Temporarily copy pasted from the execute lambda...
interface ExecuteResult {
  status: "success";
  stdout: string | null;
  stderr: string | null;
  exitCode: number | null;
  exitSignal: string | null;
  processError: string | null;
}

// Note: for problem submission, some flags to keep in mind: http://usaco.org/index.php?page=instructions

export const lambdaHandler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  const requestData = JSON.parse(event.body || "{}");
  let body;

  switch (event.httpMethod) {
    case 'POST /submissions': // create a new problem submission (POST)
      const submissionID = uuidv4();
      // todo validate structure of body?


      const params =  {
        Bucket:'cpi-onlinejudge',
        Delimiter: '/',
        Prefix: requestData.problemid + '/'
      };

      // store expected in and put in a map
        // iterate through map elemntes and run a execute function and process/store the result in an  array
        // return that array to the fromtend...

      let outFiles:string [] = [];
      let inFiles :string [] = [];
      let len:number = 0;

      const command =  new ListObjectsCommand(params);
      const data = await s3.send(command);

      const fileDetails:any = data.Contents;
      len = fileDetails.length;
      console.log(len)

      for (const file of fileDetails) {
        const fileParams = {
          Bucket: 'cpi-onlinejudge',
          Key: file.key // name of the file
        }
        const fileCommand = new GetObjectCommand(fileParams);
        const response = await s3.send(fileCommand);
        let rawFileContent = response.Body.toString('utf-8')// get file content

        const dot: number = file.key.indexOf('.');
        const tcNum: number = Number(file.key.substr(0, dot)); // get the tc number as specfied by the fileName.
        file.key.indexOf('out') !== -1 ? outFiles[tcNum] = rawFileContent : inFiles[tcNum] = rawFileContent; // match the file content to the correct index
      }

      let results: string[] = []; // execute each test case
      let status: string[] = []; // execute each test case
      for (let i = 1; i <= len; i++) {
        const execData = await execute(requestData, event, inFiles[i]);
          results[i] = execData.body;
          const output:string = JSON.parse(results[i]).body.stdout;
          if(output == outFiles[i]){
            status[i] = "correct";
          }else{
            status[i] = "wrong";
          }
      }

      const dbParams = {
      TableName: "statusTable",
          Item: {
        submissionId: submissionID,
            results: JSON.stringify(results),
            verdict: status,  //  one string result for each case ie ("compile_error")
      }
    }
      const dbCommand = new PutItemCommand(dbParams);
      const response = await db.send(dbCommand);

      return {
        statusCode: 500,
        headers: {
          'Content-Type': 'application/json',
          "Access-Control-Allow-Origin": "*",
        },
        body: JSON.stringify({
          submissionId: submissionID
        })
      };
      break;

    case 'POST /execute':

      return execute(requestData, event,  requestData.input);

      break;
    case 'GET /submissions/{submissionId}':
      const dbGetParams = {
        TableName: "statusTable",
        Key: {
          id: event.pathParameters!.submissionid
        }
      }

      const getCommand = new GetItemCommand(dbGetParams);
      const getResponse = await client.send(getCommand);

      return {
        statusCode: 500,
        headers: {
          'Content-Type': 'application/json',
          "Access-Control-Allow-Origin": "*",
        },
        body: JSON.stringify({
            status: getResponse
        })
      };

      break;
    default:
      return {
        statusCode: 100,
        headers: {
          'Content-Type': 'application/json',
          "Access-Control-Allow-Origin": "*",
        },
        body: JSON.stringify({
            status: "Inavalid event"
        })
      };
  }

};

const execute = async (requestData:any, event:APIGatewayProxyEvent, input:string) => {
  if (!requestData.language) {
    return {
      statusCode: 400,
      headers: {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
      },
      body: JSON.stringify({
        status: "Unknown language.",
      }),
    };
  }

  const compileCommand = new InvokeCommand({
    FunctionName: "online-judge-ExecuteFunction",
    Payload: Buffer.from(
        JSON.stringify({
          type: "compile",
          language: requestData.language,
          compilerOptions: requestData.compilerOptions,
          filename: requestData.filename,
          sourceCode: requestData.sourceCode,
        }),
        "utf-8"
    ),
  });
  const compileResponse = await client.send(compileCommand);
  const compileData = JSON.parse(
      Buffer.from(compileResponse.Payload!).toString()
  );

  if (compileData.status === "compile_error") {
    return {
      statusCode: 200,
      headers: {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
      },
      body: Buffer.from(compileResponse.Payload!).toString(),
    };
  } else if (compileData.status === "internal_error") {
    return {
      statusCode: 500,
      headers: {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
      },
      body: JSON.stringify({
        status: "internal_error",
        message: "Compilation failed with an internal error",
        debugData: compileData,
      }),
    };
  } else if (compileData.status !== "success") {
    return {
      statusCode: 500,
      headers: {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
      },
      body: JSON.stringify({
        status: "internal_error",
        message: "Compilation failed with an unknown error",
        debugData: compileData,
      }),
    };
  }

  const executeCommand = new InvokeCommand({
    FunctionName: "online-judge-ExecuteFunction",
    Payload: Buffer.from(
        JSON.stringify({
          type: "execute",
          payload: compileData.output,
          input:  input,
          timeout: 5000,
        })
    ),
  });
  const executeResponse = await client.send(executeCommand);
  const executeData: ExecuteResult = JSON.parse(
      Buffer.from(executeResponse.Payload!).toString()
  );

  if (executeData.status !== "success") {
    return {
      statusCode: 500,
      headers: {
        "Content-Type": "application/json",
        "Access-Control-Allow-Origin": "*",
      },
      body: JSON.stringify({
        status: "internal_error",
        message: "Execution failed for an unknown reason",
        debugData: executeData,
      }),
    };
  }

  const {stdout, exitCode, exitSignal, processError} = executeData;

  const {
    restOfString: stderr,
    time,
    memory,
  } = extractTimingInfo(executeData.stderr);




  // 124 is the exit code returned by linux `timeout` command
  if (exitCode === 124) {
    let statusList = [{
      status: "time_limit_exceeded", // get the test cases from S3 bucket;
      // TODO: Edit execute function such that it returns an array of test cases
      stdout: stdout,
      stderr: stderr,
      time: time,
      memory: memory,
      exitCode: exitCode,
    }]

    return buildResponse({
      status: "time_limit_exceeded",
      stdout,
      stderr,
      time,
      memory,
      exitCode,
    });
  }

  if (exitCode !== 0) {
    return buildResponse({
      status: "runtime_error",
      stdout,
      stderr,
      time,
      memory,
      exitCode,
    });
  }

  return buildResponse({
    status: "success",
    stdout,
    stderr,
    time,
    memory,
  });
}
