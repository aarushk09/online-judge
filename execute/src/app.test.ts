import * as app from "./app";
import compileEvent from "../events/compile.json";

test("compiles and runs C++ properly", async () => {
  const compileResult = await app.lambdaHandler(
    compileEvent as app.ExecuteEvent
  );
  expect(compileResult.status).toBe("success");

  const runResult = await app.lambdaHandler({
    type: "execute",
    payload: compileResult.output,
    input: "1 2 3",
  });
  expect(runResult).toMatchSnapshot();
});
