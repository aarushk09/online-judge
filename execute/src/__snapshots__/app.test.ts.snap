// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiles and runs C++ properly 1`] = `
Object {
  "status": "success",
  "stderr": "",
  "stdout": "6
",
}
`;
